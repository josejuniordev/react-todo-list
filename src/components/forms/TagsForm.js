import React, { Fragment, useState } from 'react';
import PropTypes from 'prop-types';
import TagClass from '../../classes/Tag';
import { Icon, Input, Tag, Tooltip } from 'antd';
import { delay } from '../../utility/Utils';

let inputRef = React.createRef();

function TagsForm(
  {
    tags = [],
    lockTheFirstTag = true,
    onTagClose = () => {},
    onInsertTag = () => {},
  }
) {

  const [isInputVisible, setIsInputVisible] = useState(false);
  const [inputValue, setInputValue] = useState("");

  function handleInputChange(ev) {
    setInputValue(ev.target.value);
  }

  function handleInputConfirm() {
    const isTagAlreadyInserted = tags.find(tag => tag.name === inputValue);

    if (!isTagAlreadyInserted) {
      onInsertTag(inputValue);
    }

    setIsInputVisible(false);
    setInputValue('');
  }

  function handleNewTagClick() {
    setIsInputVisible(true);

    delay(200).then(() => {
      if (inputRef.current) {
        inputRef.current.focus();
      }
    });
  }

  return (
    <Fragment>
      {
        tags.map((tag, index) => {
          const isLongTag = tag.length > 20;
          let isClosable = index !== 0;

          if (!lockTheFirstTag) {
            isClosable = true;
          }

          const tagElement = (
            <Tag
              key={tag.id}
              closable={isClosable}
              afterClose={() => onTagClose(tag.id)}
            >
              {isLongTag ? `${tag.name.slice(0, 20)}...` : tag.name}
            </Tag>
          );
          return isLongTag ? <Tooltip title={tag.name} key={tag.id}>{tagElement}</Tooltip> : tagElement;
        })
      }

      {isInputVisible && (
        <Input
          ref={inputRef}
          type="text"
          size="small"
          style={{ width: 78 }}
          value={inputValue}
          onChange={handleInputChange}
          onBlur={handleInputConfirm}
          onPressEnter={handleInputConfirm}
        />
      )}

      {!isInputVisible && (
        <Tag
          onClick={handleNewTagClick}
          style={{ background: '#fff', borderStyle: 'dashed' }}
        >
          <Icon type="plus" /> Nova Tag
        </Tag>
      )}
    </Fragment>
  )
}

TagsForm.propTypes = {
  tags: PropTypes.arrayOf(PropTypes.instanceOf(TagClass)).isRequired,
  lockTheFirstTag: PropTypes.bool,
  onTagClose: PropTypes.func,
  onInsertTag: PropTypes.func,
};

export default TagsForm;
