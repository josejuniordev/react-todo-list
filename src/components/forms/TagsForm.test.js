import React from 'react';
import { mount, shallow } from 'enzyme';
import TagsForm from './TagsForm';
import tagsMockedData from '../../mock/TagsMock';
import { Tag } from 'antd';

describe('TagsForm component', () => {
  const rendered = shallow(<TagsForm tags={tagsMockedData} />);

  test('should render correctly', () => {
    expect(rendered).toMatchSnapshot();
  });

  test('renders the correct number of tags', () => {
    const tags = rendered.find(Tag);
    const tagsLengthPlusOneStatic = tagsMockedData.length + 1;

    expect(tags.length).toBe(tagsLengthPlusOneStatic);
  });

});
