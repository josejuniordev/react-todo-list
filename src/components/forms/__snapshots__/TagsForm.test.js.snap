// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagsForm component should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TagsForm
    tags={
      Array [
        Tag {
          "createdAt": "2019-03-22T01:45:05.731Z",
          "id": 1,
          "name": "Estudo",
        },
        Tag {
          "createdAt": "2019-03-21T03:55:15.599Z",
          "id": 2,
          "name": "Trabalho",
        },
        Tag {
          "createdAt": "2019-03-21T18:49:02.309Z",
          "id": 3,
          "name": "Teste",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        Array [
          <Tag
            afterClose={[Function]}
            closable={false}
          >
            Estudo
          </Tag>,
          <Tag
            afterClose={[Function]}
            closable={true}
          >
            Trabalho
          </Tag>,
          <Tag
            afterClose={[Function]}
            closable={true}
          >
            Teste
          </Tag>,
        ],
        false,
        <Tag
          closable={false}
          onClick={[Function]}
          style={
            Object {
              "background": "#fff",
              "borderStyle": "dashed",
            }
          }
        >
          <Icon
            type="plus"
          />
           Nova Tag
        </Tag>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "class",
        "props": Object {
          "afterClose": [Function],
          "children": "Estudo",
          "closable": false,
        },
        "ref": null,
        "rendered": "Estudo",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "class",
        "props": Object {
          "afterClose": [Function],
          "children": "Trabalho",
          "closable": true,
        },
        "ref": null,
        "rendered": "Trabalho",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "3",
        "nodeType": "class",
        "props": Object {
          "afterClose": [Function],
          "children": "Teste",
          "closable": true,
        },
        "ref": null,
        "rendered": "Teste",
        "type": [Function],
      },
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Icon
              type="plus"
            />,
            " Nova Tag",
          ],
          "closable": false,
          "onClick": [Function],
          "style": Object {
            "background": "#fff",
            "borderStyle": "dashed",
          },
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "type": "plus",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          " Nova Tag",
        ],
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          Array [
            <Tag
              afterClose={[Function]}
              closable={false}
            >
              Estudo
            </Tag>,
            <Tag
              afterClose={[Function]}
              closable={true}
            >
              Trabalho
            </Tag>,
            <Tag
              afterClose={[Function]}
              closable={true}
            >
              Teste
            </Tag>,
          ],
          false,
          <Tag
            closable={false}
            onClick={[Function]}
            style={
              Object {
                "background": "#fff",
                "borderStyle": "dashed",
              }
            }
          >
            <Icon
              type="plus"
            />
             Nova Tag
          </Tag>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "class",
          "props": Object {
            "afterClose": [Function],
            "children": "Estudo",
            "closable": false,
          },
          "ref": null,
          "rendered": "Estudo",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "class",
          "props": Object {
            "afterClose": [Function],
            "children": "Trabalho",
            "closable": true,
          },
          "ref": null,
          "rendered": "Trabalho",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "3",
          "nodeType": "class",
          "props": Object {
            "afterClose": [Function],
            "children": "Teste",
            "closable": true,
          },
          "ref": null,
          "rendered": "Teste",
          "type": [Function],
        },
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Icon
                type="plus"
              />,
              " Nova Tag",
            ],
            "closable": false,
            "onClick": [Function],
            "style": Object {
              "background": "#fff",
              "borderStyle": "dashed",
            },
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "type": "plus",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            " Nova Tag",
          ],
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
